core/src/cell.rs

```diff
@@ -70,39 +70,67 @@ impl CellStatus {
     pub fn is_unknown(&self) -> bool {
         self == &CellStatus::Unknown
     }
-
-    pub fn get_live_output(&self) -> Option<&CellMeta> {
-        match *self {
-            CellStatus::Live(LiveCell::Output(ref output)) => Some(output),
-            _ => None,
-        }
-    }
-
-    pub fn take_live_output(self) -> Option<CellMeta> {
-        match self {
-            CellStatus::Live(LiveCell::Output(output)) => Some(output),
-            _ => None,
-        }
-    }
 }
 
 /// Transaction with resolved input cells.
 #[derive(Debug)]
-pub struct ResolvedTransaction {
-    pub transaction: Transaction,
-    pub dep_cells: Vec<CellStatus>,
-    pub input_cells: Vec<CellStatus>,
+pub struct ResolvedTransaction<'a> {
+    pub transaction: &'a Transaction,
+    pub dep_cells: Vec<CellMeta>,
+    pub input_cells: Vec<CellMeta>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum UnresolvableError {
+    Dead(OutPoint),
+    Unknown(OutPoint),
 }
 
 pub trait CellProvider {
     fn cell(&self, out_point: &OutPoint) -> CellStatus;
 
-    fn get_cell_status(&self, out_point: &OutPoint) -> CellStatus {
-        if out_point.is_null() {
-            CellStatus::Live(LiveCell::Null)
+    fn resolve_transaction<'a>(
```

Comment:
Please keep `resolve_transaction` as a separate function outside of this. The problem is that in order to generalize OutPoint, we will add a separate HeaderProvider to `resolve_transaction` so it can also resolve headers. Putting `resolve_transaction` here makes it hard to deal with HeaderProvider, and we might still need to extract `resolve_transaction` out of CellProvider.

Reply:
We can add a helper fn for `HeaderProvider` if needed:

```rust
fn resolve_transaction(tx: &Transaction, cp: &CellProvider, hp: &HeaderProvider) -> ResolvedTransaction {
    cp.resolve_transaction(tx).and_then(|hp| hp.resolve_transaction(tx))
}
```

Or a `TransactionResolver` trait/struct like this:

```rust
struct TransactionResolver {
    cp: &CellProvider,
    hp: &HeaderProvider,
}

impl TransactionResolver {
    fn resolve(tx: &Transaction) {
        
    }
}
```

Reply:
Personally I don't see how this is better than a plain function and the need to put `resolve_transaction` function within CellProvider trait, but it's not a big deal here, I don't have strong feelings.

Reply:
Before this PR, `resolve_transaction` fn needs a mutable `seen_inputs` map as the parameter, this PR removes it, and all the codes are related to `CellProvider` inner logic now, so I encapsulate them to `CellProvider`.

Reply:
`BlockCellProvider` is highly coupled with the block verification logic in `ChainService::reconcile_main_chain`. This is really a problem. `BlockCellProvider` is not a cell set status after applied the transactions in Block, instead, it is partially applied. And the partially applied status just happens to be working in the block verification.

Reply:
At least, `BlockCellProvider` should be renamed, and it should reside inside crate `chain`. It does not look a struct can be shared, it is designed dedicated to the block verification. So does `TransactionCellProvider`.